File: 3.x.x/crates-utopia/cache/src/adapter/pool.rs
use std::sync::Arc;

use utopia_pools::{Pool as UtopiaPool, Recover, RecoverCall, Stack};

use crate::adapter::Adapter;
use crate::error::CacheError;
use crate::feature::Leasable;
use crate::value::{CacheValue, LoadResult, SaveResult};

/// Local wrapper so [`Recover`] can be implemented for pooled adapters.
#[derive(Clone)]
pub struct PooledAdapter(pub Arc<dyn Adapter>);

impl std::fmt::Debug for PooledAdapter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("PooledAdapter").finish()
    }
}

impl Recover for PooledAdapter {
    fn reset(&mut self) -> RecoverCall {
        RecoverCall::Succeeded
    }
}

impl Adapter for PooledAdapter {
    fn load(&self, key: &str, ttl: i64, hash: &str) -> Result<LoadResult, CacheError> {
        self.0.load(key, ttl, hash)
    }

    fn save(&self, key: &str, data: &CacheValue, hash: &str) -> Result<SaveResult, CacheError> {
        self.0.save(key, data, hash)
    }

    fn touch(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        self.0.touch(key, hash)
    }

    fn list(&self, key: &str) -> Result<Vec<String>, CacheError> {
        self.0.list(key)
    }

    fn purge(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        self.0.purge(key, hash)
    }

    fn flush(&self) -> Result<bool, CacheError> {
        self.0.flush()
    }

    fn ping(&self) -> bool {
        self.0.ping()
    }

    fn get_size(&self) -> Result<i64, CacheError> {
        self.0.get_size()
    }

    fn get_name(&self, key: Option<&str>) -> String {
        self.0.get_name(key)
    }

    fn as_leasable(&self) -> Option<&dyn Leasable> {
        self.0.as_leasable()
    }
}

/// PHP `Utopia\Pools\Pool` of cache adapters, used by unit tests and [`Pool`].
pub type AdapterPool = UtopiaPool<PooledAdapter>;

/// Builds a [`utopia_pools::Pool`] wrapping one adapter instance (Arc).
///
/// PHP tests construct `new Utopia\Pools\Pool(new Stack(), …, fn(): Adapter => …)`.
#[derive(Debug)]
pub struct MemoryPool;

impl MemoryPool {
    #[must_use]
    pub fn single(adapter: impl Adapter + 'static) -> AdapterPool {
        let shared: Arc<dyn Adapter> = Arc::new(adapter);
        UtopiaPool::new(
            Stack::new(),
            "test",
            1,
            move || PooledAdapter(Arc::clone(&shared)),
            0.0,
        )
        .expect("cache adapter pool")
    }
}

/// PHP `Utopia\Cache\Adapter\Pool`.
pub struct Pool {
    pool: AdapterPool,
}

impl std::fmt::Debug for Pool {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Pool").finish_non_exhaustive()
    }
}

impl Pool {
    /// PHP `__construct(Pool $pool)`. Validates the pool holds adapters.
    pub fn new(pool: AdapterPool) -> Result<Self, CacheError> {
        pool.use_sync(|adapter| adapter.get_name(None))
            .map_err(|err| CacheError::message(err.to_string()))?;
        Ok(Self { pool })
    }

    fn delegate<R>(
        &self,
        f: impl FnOnce(&PooledAdapter) -> Result<R, CacheError>,
    ) -> Result<R, CacheError> {
        match self.pool.use_sync(|adapter| f(adapter)) {
            Ok(result) => result,
            Err(err) => Err(CacheError::message(err.to_string())),
        }
    }
}

impl Adapter for Pool {
    fn load(&self, key: &str, ttl: i64, hash: &str) -> Result<LoadResult, CacheError> {
        self.delegate(|a| a.load(key, ttl, hash))
    }

    fn save(&self, key: &str, data: &CacheValue, hash: &str) -> Result<SaveResult, CacheError> {
        self.delegate(|a| a.save(key, data, hash))
    }

    fn touch(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        self.delegate(|a| a.touch(key, hash))
    }

    fn list(&self, key: &str) -> Result<Vec<String>, CacheError> {
        self.delegate(|a| a.list(key))
    }

    fn purge(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        self.delegate(|a| a.purge(key, hash))
    }

    fn flush(&self) -> Result<bool, CacheError> {
        self.delegate(|a| a.flush())
    }

    fn ping(&self) -> bool {
        self.pool.use_sync(|a| a.ping()).unwrap_or(false)
    }

    fn get_size(&self) -> Result<i64, CacheError> {
        self.delegate(|a| a.get_size())
    }

    fn get_name(&self, key: Option<&str>) -> String {
        self.pool
            .use_sync(|a| a.get_name(key))
            .unwrap_or_else(|_| "pool".into())
    }

    fn as_leasable(&self) -> Option<&dyn Leasable> {
        Some(self)
    }
}

impl Leasable for Pool {
    fn get_generation(&self, key: &str) -> Result<String, CacheError> {
        self.delegate(|a| match a.as_leasable() {
            Some(leasable) => leasable.get_generation(key),
            None => Ok("0".into()),
        })
    }

    fn save_with_lease(
        &self,
        key: &str,
        data: &CacheValue,
        hash: &str,
        generation: &str,
    ) -> Result<SaveResult, CacheError> {
        self.delegate(|a| match a.as_leasable() {
            Some(leasable) => leasable.save_with_lease(key, data, hash, generation),
            None => a.save(key, data, hash),
        })
    }
}

File: packages/cache/src/Cache/Adapter/Pool.php
<?php

namespace Utopia\Cache\Adapter;

use Utopia\Cache\Adapter;
use Utopia\Cache\Feature\Batchable;
use Utopia\Cache\Feature\Leasable;
use Utopia\Pools\Pool as UtopiaPool;

class Pool implements Adapter, Batchable, Leasable
{
    /**
     * @param  UtopiaPool<covariant Adapter>  $pool The pool to use for connections. Must contain instances of Adapter.
     *
     * @throws \Exception
     */
    public function __construct(protected UtopiaPool $pool)
    {
        $this->pool->use(function (mixed $resource): void {
            if (! ($resource instanceof Adapter)) {
                throw new \Exception('Pool must contain instances of ' . Adapter::class);
            }
        });
    }

    /**
     * Forward method calls to the internal adapter instance via the pool.
     *
     * Required because __call() can't be used to implement abstract methods.
     *
     * @param  array<mixed>  $args
     */
    public function delegate(string $method, array $args): mixed
    {
        return $this->pool->use(fn(Adapter $adapter) => $adapter->{$method}(...$args));
    }

    public function load(string $key, int $ttl, string $hash = ''): mixed
    {
        return $this->delegate(__FUNCTION__, \func_get_args());
    }

    public function save(string $key, array|string $data, string $hash = '', int $ttl = 0): bool|string|array
    {
        /**
         * @var bool|string|array<mixed> $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    /**
     * @param  string[]  $fields
     * @param  int  $ttl time in seconds
     * @return array<string, mixed>
     */
    public function loadMany(string $key, array $fields, int $ttl): array
    {
        /** @var array<string, mixed> $result */
        $result = $this->pool->use(fn(Adapter $adapter): array => $adapter instanceof Batchable ? $adapter->loadMany($key, $fields, $ttl) : []);

        return $result;
    }

    /**
     * @param  array<string, mixed>  $data field => value
     * @param  int  $ttl time in seconds
     * @return array<string, mixed>|false
     */
    public function saveMany(string $key, array $data, int $ttl = 0): array|false
    {
        /** @var array<string, mixed>|false $result */
        $result = $this->pool->use(fn(Adapter $adapter): array|false => $adapter instanceof Batchable ? $adapter->saveMany($key, $data, $ttl) : false);

        return $result;
    }

    public function getGeneration(string $key): string
    {
        return $this->pool->use(fn(Adapter $adapter): string => $adapter instanceof Leasable ? $adapter->getGeneration($key) : '0');
    }

    public function saveWithLease(string $key, array|string $data, string $hash, string $generation): bool|string|array
    {
        /**
         * @var bool|string|array<mixed> $result
         */
        $result = $this->pool->use(fn(Adapter $adapter): bool|string|array => $adapter instanceof Leasable
            ? $adapter->saveWithLease($key, $data, $hash, $generation)
            : $adapter->save($key, $data, $hash));

        return $result;
    }

    public function touch(string $key, string $hash = ''): bool
    {
        /**
         * @var bool $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    public function list(string $key): array
    {
        /**
         * @var array<string> $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    public function purge(string $key, string $hash = ''): bool
    {
        /**
         * @var bool $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    public function flush(): bool
    {
        /**
         * @var bool $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    public function ping(): bool
    {
        /**
         * @var bool $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    public function getSize(): int
    {
        /**
         * @var int $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }

    public function getName(?string $key = null): string
    {
        /**
         * @var string $result
         */
        $result = $this->delegate(__FUNCTION__, \func_get_args());

        return $result;
    }
}
